### DISCUSSION TITLE
int i = 0; vs int i {0};

### DISCUSSION THREAD

**Discussion by H1manshu21** [Blaze-Stars/Image-Filter-Application#18] [Q&A] [answered]
- `int i = 0` means the compiler will store a `garbage value` first then it will assign `0` to `i`.
- `int i {0}` means the compiler will store `0` directly without initializing to a `garbage value`.

This reduces two cycles to one cycle by using `{}`.
This is introduced in `C++11`.

**Comment by SHembram18** [comment]
▪ But generally why we use `int i = 0` instead of `int i {0}`🤔
@Himanshu40

**Comment by H1manshu21** [comment, marked as answer]
`int i = 0;` is an old style of initializing variables in C and C++. From `C++11` onwards, `initialization` and `assignment` are different terms. Initialization is more effective for class member variables.
Must read: https://www.learncpp.com/cpp-tutorial/variable-assignment-and-initialization/#:~:text=Variable%20assignment%20and%20initialization

**Comment by H1manshu21** [comment]
### What is the difference between Initializing And Assignment inside the constructor? & What is the advantage?

- There is a difference between Initializing a member using an initializer list and assigning it a value inside the constructor body.

- When you initialize fields via the initializer list the constructors will be called once.

- If you use the assignment then the fields will be first initialized with default constructors and then reassigned (via assignment operator) with actual values.

As you see there is an additional overhead of creation & assignment in the latter, which might be considered for user-defined classes.
For an integer data type or POD class members there is no practical overhead.

**An Code Example:**
```cpp
class Myclass {
public: 
    Myclass (unsigned int param) : param_ (param) {}
    
    unsigned int param () const {
        return param_;
     }

private:
     unsigned int param_;
};
```
**In the above example:**

`Myclass (unsigned int param) : param_ (param)`
This construct is called a `Member Initializer List` in C++.

It initializes a member `param_` to a value `param`.

### When do you HAVE TO use the member Initializer list?
You will have(rather forced) to use a Member Initializer list if:

- Your class has a reference member
- Your class has a const member or
- Your class doesn't have a default constructor

**Comment by H1manshu21** [comment]
### What is POD?
Must read: https://stackoverflow.com/questions/146452/what-are-pod-types-in-c#:~:text=pointers%20(including%20pointer%2Dto%2D,%2C%20destructors%2C%20or%20virtual%20methods.
